File: core/src/main/java/co/fastestpath/api/schedule/ServiceId.java
package co.fastestpath.api.schedule;

public class ServiceId {

  private final String value;

  public ServiceId(String value) {
    this.value = value;
  }

  @Override
  public String toString() {
    return value;
  }

  @Override
  public boolean equals(Object o) {
    if (this == o) return true;
    if (o == null || getClass() != o.getClass()) return false;

    ServiceId stopId = (ServiceId) o;

    return value != null ? value.equals(stopId.value) : stopId.value == null;
  }

  @Override
  public int hashCode() {
    return value != null ? value.hashCode() : 0;
  }
}

File: core/src/main/kotlin/co/fastestpath/api/schedule/ServiceId.kt
package co.fastestpath.api.schedule

data class ServiceId private constructor(val value: String) {
  companion object {

    private val cache: MutableMap<String, ServiceId> = HashMap()

    fun of(value: String): ServiceId {
      var serviceId = cache[value]
      if (serviceId != null) {
        return serviceId
      }
      serviceId = ServiceId(value)
      cache[value] = serviceId
      return serviceId
    }
  }
}
